cli/src/lib/compiler/compile.js

```diff
@@ -1,17 +1,140 @@
+        const msg = `Could not resolve app entrypoint ${chalk.bold(entrypoint)}`
+        reporter.error(msg)
+        throw new Error(msg)
+    }
+
+    if (isApp) {
+        const shellAppSource = await fs.readFile(paths.shellSourceEntrypoint)
+        await fs.writeFile(
+            paths.shellAppEntrypoint,
+            shellAppSource
+                .toString()
+                .replace(/'.\/D2App\/app'/g, `'./D2App/${relativeEntrypoint}'`)
+        )
+    }
+}
+
+const watchFiles = ({ inputDir, outputDir, processFileCallback, watch }) => {
+    const compileFile = async source => {
+        const relative = path.relative(inputDir, source)
+        const destination = path.join(outputDir, relative)
+        reporter.debug(
+            `File ${relative} changed or added... dest: `,
+            path.relative(inputDir, relative)
+        )
+        await fs.ensureDir(path.dirname(destination))
+        await processFileCallback(source, destination)
+    }
+
+    const removeFile = async file => {
+        const relative = path.relative(inputDir, file)
+        const outFile = path.join(outputDir, relative)
+        reporter.debug(`File ${relative} removed... removing: `, outFile)
+        fs.remove(outFile)
+    }
+
+    return new Promise((resolve, reject) => {
+        const watcher = chokidar.watch(inputDir, { persistent: true })
+
+        watcher
+            .on('ready', async () => {
+                if (watch) {
+                    reporter.debug('watching...')
+                } else {
+                    await watcher.close()
+                }
+                resolve()
+            })
+            .on('add', compileFile)
+            .on('change', compileFile)
+            .on('unlink', removeFile)
+            .on('error', error => {
+                reporter.debugErr('Chokidar error:', error)
+                reject('Chokidar error!')
+            })
+
+        process.on('SIGINT', async () => {
+            reporter.debug('Caught interrupt signal')
+            await watcher.close()
+        })
+    })
+}
 
 const compile = async ({
     config,
     paths,
+    moduleType = 'es',
     mode = 'development',
     watch = false,
-} = {}) => {
-    if (config.type === 'lib') {
-        return await compileLibrary({ config, paths, mode, watch })
-    } else {
-        return await compileApp({ config, paths, mode, watch })
+}) => {
+    await overwriteEntrypoint({ config, paths })
+
+    const isApp = config.type === 'app'
+    const outDir = isApp
+        ? paths.shellApp
+        : path.join(paths.buildOutput, moduleType)
+
+    fs.removeSync(outDir)
+    fs.ensureDirSync(outDir)
+
+    if (isApp) {
+        fs.removeSync(paths.shellPublic)
+        fs.copySync(paths.shellSourcePublic, paths.shellPublic)
+    }
+
+    const copyFile = async (source, destination) => {
+        await fs.copy(source, destination)
     }
+    const compileFile = async (source, destination) => {
+        if (path.extname(source) === '.js') {
```

Comment:
Should we also check if the `extname` is `.jsx` or `.ts`?

Reply:
Yep, nice catch!  Fixed